src/components/util/checkbox-list.js

```diff
@@ -4,55 +4,111 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
-import React, { useState } from 'react';
+import React, { useEffect, useState } from 'react';
 import List from '@material-ui/core/List';
 import ListItem from '@material-ui/core/ListItem';
 import ListItemIcon from '@material-ui/core/ListItemIcon';
 import ListItemText from '@material-ui/core/ListItemText';
 import Checkbox from '@material-ui/core/Checkbox';
+import DeleteIcon from '@material-ui/icons/Delete';
+import IconButton from '@material-ui/core/IconButton';
+import * as PropTypes from 'prop-types';
 
-const CheckboxList = (props) => {
-    const [checked, setChecked] = useState([]);
+function CustomListItem(props) {
+    const [isHoover, setHoover] = useState(false);
+
+    function handleHoover(enter) {
+        return setHoover(enter);
+    }
+
+    return (
+        <ListItem
+            role={undefined}
+            dense
+            button
+            onClick={props.onClick}
+            onMouseEnter={() => handleHoover(true)}
+            onMouseLeave={() => handleHoover(false)}
+        >
+            <ListItemIcon>
+                <Checkbox
+                    color={'primary'}
+                    edge="start"
+                    checked={props.set.has(props.value)}
+                    tabIndex={-1}
+                    disableRipple
+                />
+            </ListItemIcon>
+            <ListItemText primary={props.primary} />
+            {props.removeFromList && isHoover && (
+                <IconButton onClick={props.removeFromList} size={'small'}>
+                    <DeleteIcon
+                        style={{
+                            alignItems: 'end',
+                        }}
+                    />
+                </IconButton>
+            )}
+        </ListItem>
+    );
+}
 
-    const handleToggle = (value) => () => {
-        const currentIndex = checked.indexOf(value);
-        const newChecked = [...checked];
+CustomListItem.propTypes = {
+    onClick: PropTypes.func,
+    set: PropTypes.any,
+    value: PropTypes.any,
+    primary: PropTypes.any,
+    removeFromList: PropTypes.any,
+};
+const CheckboxList = (props) => {
+    const [checked, setChecked] = useState(new Set(props.initialSelection));
 
-        if (currentIndex === -1) {
-            newChecked.push(value);
-        } else {
-            newChecked.splice(currentIndex, 1);
+    /* remove non absent selected items */
+    useEffect(() => {
+        const existingValues = new Set(props.values.map(props.id));
+        const newChecked = new Set(
+            [...checked].filter((id) => existingValues.has(id))
+        );
+        if (newChecked.size !== checked.size) {
```

Comment:
This is bad, you could have add a value and remove one other then the sizes are equals but checked must be updated

Reply:
nope, the new elements must exists in the old set, if the size are equals then the set are equals